utils/s2n_blob.c

```diff
@@ -50,16 +50,21 @@ int s2n_blob_zero(struct s2n_blob *b)
 
 int s2n_blob_slice(const struct s2n_blob *b, struct s2n_blob *slice, uint32_t offset, uint32_t size)
 {
-    notnull_check(b);
-    notnull_check(slice);
-
-    S2N_ERROR_IF(b->size < (offset + size), S2N_ERR_SIZE_MISMATCH);
-    slice->data = b->data + offset;
-    slice->size = size;
-    slice->growable = 0;
-    slice->allocated = 0;
-
-    return 0;
+    PRECONDITION_POSIX(s2n_blob_is_valid(b));
+    PRECONDITION_POSIX(s2n_blob_is_valid(slice));
```

Comment:
This is technically changing the error returned, which will be less obvious about why it failed: "precondition violation" vs "null pointer error". I'm wondering if it would be beneficial, instead, have a `int s2n_blob_validate(const struct s2n_blob *blob)` function that we can just `GUARD` here:

```c
int s2n_blob_validate(const struct s2n_blob* b)
{
    ENSURE_POSIX_REF(b);
    ENSURE_POSIX(S2N_IMPLIES(b->growable == 0, b->allocated == 0), ERROR_CODE);
    ENSURE_POSIX(S2N_IMPLIES(b->growable != 0, S2N_MEM_IS_READABLE(b->data, b->allocated)), ERROR_CODE);
    ENSURE_POSIX(S2N_IMPLIES(b->growable != 0, b->size <= b->allocated), ERROR_CODE);
    ENSURE_POSIX(S2N_MEM_IS_READABLE(b->data, b->size), ERROR_CODE);
    return S2N_SUCCESS;
}

int s2n_blob_slice(const struct s2n_blob *b, struct s2n_blob *slice, uint32_t offset, uint32_t size)
{
    GUARD(s2n_blob_validate(b));
    GUARD(s2n_blob_validate(slice));
    ...
    GUARD(s2n_blob_validate(slice));
}
```

With some new fancy safety macros, which would replace all the _NONNULL, which i never liked anyway as it's inverted logic.

```c
// s2n_safety.h


/**
 * Ensures `x` is a readable reference, otherwise the function will `BAIL` with `S2N_ERR_NULL`
 */
#define ENSURE_REF( x )                              ENSURE(S2N_OBJECT_PTR_IS_READABLE(x), S2N_ERR_NULL)

/**
 * Ensures `x` is a readable reference, otherwise the function will `BAIL_POSIX` with `S2N_ERR_NULL`
 */
#define ENSURE_POSIX_REF( x )                        ENSURE_POSIX(S2N_OBJECT_PTR_IS_READABLE(x), S2N_ERR_NULL)

/**
 * Ensures `x` is a mutable reference, otherwise the function will `BAIL` with `S2N_ERR_NULL`
 */
#define ENSURE_MUT( x )                              ENSURE(S2N_OBJECT_PTR_IS_WRITABLE(x), S2N_ERR_NULL)

/**
 * Ensures `x` is a mutable reference, otherwise the function will `BAIL_POSIX` with `S2N_ERR_NULL`
 */
#define ENSURE_POSIX_MUT( x )                        ENSURE_POSIX(S2N_OBJECT_PTR_IS_WRITABLE(x), S2N_ERR_NULL)
```

Reply:
That's a great suggestion. It belongs to its own PR, so I moved the discussion to PR https://github.com/awslabs/s2n/pull/1990.